tests/location.rs

```diff
@@ -0,0 +1,476 @@
+        while let Some(item) = queue.pop_front() {
+            self.log_event(&item);
+        }
+    }
+
+    fn log_event(&self, item: &EventsItem) {
+        match &item.event {
+            Event::Info(msg) => println!("I [{} {:?}]: {}", item.acc_name, item.when, msg),
+            Event::Warning(msg) => println!("W [{} {:?}]: {}", item.acc_name, item.when, msg),
+            Event::Error(msg) => println!("E [{} {:?}]: {}", item.acc_name, item.when, msg),
+            _ => println!("Evt [{} {:?}]: {:?}", item.acc_name, item.when, item.event),
+        }
+    }
+
+    fn clear_log_events(&self) {
+        let mut queue = self.events.lock().unwrap();
+        for item in queue.iter() {
+            self.log_event(item)
+        }
+        queue.retain(|item| match item.event {
+            Event::Info(_) | Event::Warning(_) | Event::Error(_) => false,
+            _ => true,
+        });
+    }
+}
+
+/// A Configured DeltaChat account.
+#[derive(Debug)]
+struct Account {
+    name: String,
+    creds: AccountCredentials,
+    ctx: Arc<Context>,
+    events: Arc<EventsQueue>,
+    running: Arc<atomic::AtomicBool>,
+    imap_handle: Option<thread::JoinHandle<()>>,
+    mvbox_handle: Option<thread::JoinHandle<()>>,
+    sentbox_handle: Option<thread::JoinHandle<()>>,
+    smtp_handle: Option<thread::JoinHandle<()>>,
+}
+
+impl Account {
+    fn new(name: &str, dir: &Path, keys: KeyPair, start: std::time::Instant) -> Account {
+        // Create events queue and callback.
+        let events = Arc::new(EventsQueue::new(name));
+        let events_cb = Arc::clone(&events);
+        let name_cb = name.to_string();
+        let cb = move |_ctx: &Context, evt: Event| -> uintptr_t {
+            events_cb.push(EventsItem {
+                acc_name: name_cb.clone(),
+                when: start.elapsed(),
+                event: evt,
+            });
+            0
+        };
+
+        // Create and configure the context.
+        let dbfile = dir.join(format!("{}.db", name));
+        let creds = AccountCredentials::new(&Account::liveconfig_url());
+        println!("Account credentials for {}: {:#?}", name, creds);
+        let ctx = Arc::new(Context::new(Box::new(cb), "TestClient".into(), dbfile).unwrap());
+        ctx.set_config(Config::Addr, Some(&creds.email)).unwrap();
+        ctx.set_config(Config::MailPw, Some(&creds.password))
+            .unwrap();
+        keys.save_as_self(&ctx);
+        deltachat::configure::configure(&ctx);
+
+        // Start the threads.
+        let running = Arc::new(atomic::AtomicBool::new(true));
+        let imap_handle = Self::start_imap(name, Arc::clone(&ctx), Arc::clone(&running));
+        let mvbox_handle = Self::start_mvbox(name, Arc::clone(&ctx), Arc::clone(&running));
+        let sentbox_handle = Self::start_sentbox(name, Arc::clone(&ctx), Arc::clone(&running));
+        let smtp_handle = Self::start_smtp(name, Arc::clone(&ctx), Arc::clone(&running));
```

Comment:
@dignifiedquire, @link2xt, @KAction  These functions are entirely identical other than the names of the functions they invoke, which are easily guessable.  I tried to replace them with a macro like `start_thread!(imap, ctx, running)` or something but my macro-fu is significantly lacking.  Anyone know how to do this?